include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -357,40 +357,50 @@ struct __peer_prefix_helper<_OffsetT, __peer_prefix_algo::atomic_fetch_or>
     ::std::uint32_t __self_lidx;
     ::std::uint32_t __item_mask;
     _AtomicT __atomic_peer_mask;
+    static constexpr std::uint32_t __radix_states = 1 << __radix_bits;
 
     __peer_prefix_helper(sycl::nd_item<1> __self_item, _TempStorageT __lacc)
         : __sgroup(__self_item.get_sub_group()), __self_lidx(__self_item.get_local_linear_id()),
           __item_mask(~(~0u << (__self_lidx))), __atomic_peer_mask(*__dpl_sycl::__get_accessor_ptr(__lacc))
     {
     }
 
-    ::std::uint32_t
-    __peer_contribution(_OffsetT& __new_offset_idx, _OffsetT __offset_prefix, bool __is_current_bucket)
+    template <typename _OffsetHistogramAcc>
+    _OffsetT
+    __peer_contribution(std::uint32_t __bucket, _OffsetHistogramAcc& __histogram)
     {
-        // reset mask for each radix state
-        if (__self_lidx == 0)
-            __atomic_peer_mask.store(0U);
-        sycl::group_barrier(__sgroup);
-        // set local id's bit to 1 if the bucket value matches the radix state
-        __atomic_peer_mask.fetch_or(__is_current_bucket << __self_lidx);
-        sycl::group_barrier(__sgroup);
-        ::std::uint32_t __peer_mask_bits = __atomic_peer_mask.load();
-        ::std::uint32_t __sg_total_offset = sycl::popcount(__peer_mask_bits);
-
-        // get the local offset index from the bits set in the peer mask with index less than the work item ID
-        __peer_mask_bits &= __item_mask;
-        __new_offset_idx |= __is_current_bucket * (__offset_prefix + sycl::popcount(__peer_mask_bits));
-        return __sg_total_offset;
+        _OffsetT __offset = 0;
+        _ONEDPL_PRAGMA_UNROLL
```

Comment:
We use 8 radix bits and 256 resulting states, right? This seems like a lot to me unrolling. Did you experiment without this directive? Or with using a smaller value?

We've seen increased compile times when using unrolling directives in the past. Did you check if this has an effect here as well?

Reply:
We currently sort 4 bits per pass, but it may be more in the future. I am going to remove the pragma, to limit the PR only to such refactoring.